File: 01-machine/04-races-and-atomicity/cpp/race.cpp
// Layer 1 - Why `counter++` is not atomic, C++ version.
// Unsynchronized concurrent access to a plain (non-atomic) variable is
// undefined behavior under the C++ memory model, exactly like Rust's rule
// for non-atomic shared data -- and for the same reason Rust's bare
// increment race didn't reproduce cleanly in this lab (see that topic's
// README), the optimizer is allowed to assume it doesn't happen and
// transform the loop accordingly. C++ hands you the footgun without even
// requiring an `unsafe` keyword to pull the trigger, which is precisely
// the contrast worth noticing against Rust.
#include <atomic>
#include <cstdio>
#include <mutex>
#include <thread>
#include <vector>

constexpr int THREADS = 8;
constexpr long INCREMENTS = 300'000;

long run_unsafe() {
    long counter = 0;
    std::vector<std::thread> threads;
    for (int i = 0; i < THREADS; i++) {
        threads.emplace_back([&] {
            for (long j = 0; j < INCREMENTS; j++) {
                counter++; // racy: not synchronized, and legally UB
            }
        });
    }
    for (auto& t : threads) t.join();
    return counter;
}

long run_mutex() {
    long counter = 0;
    std::mutex m;
    std::vector<std::thread> threads;
    for (int i = 0; i < THREADS; i++) {
        threads.emplace_back([&] {
            for (long j = 0; j < INCREMENTS; j++) {
                std::lock_guard<std::mutex> lk(m);
                counter++;
            }
        });
    }
    for (auto& t : threads) t.join();
    return counter;
}

long run_atomic() {
    std::atomic<long> counter{0};
    std::vector<std::thread> threads;
    for (int i = 0; i < THREADS; i++) {
        threads.emplace_back([&] {
            for (long j = 0; j < INCREMENTS; j++) {
                counter.fetch_add(1, std::memory_order_relaxed);
            }
        });
    }
    for (auto& t : threads) t.join();
    return counter.load();
}

int main() {
    long expected = static_cast<long>(THREADS) * INCREMENTS;
    long unsafe_result = run_unsafe();
    long mutex_result = run_mutex();
    long atomic_result = run_atomic();
    std::printf("expected:               %ld\n", expected);
    std::printf("unsafe (counter++):     %ld  (lost %ld)\n", unsafe_result, expected - unsafe_result);
    std::printf("safe (std::mutex):      %ld\n", mutex_result);
    std::printf("safe (std::atomic):     %ld\n", atomic_result);
    return 0;
}

File: 01-machine/04-races-and-atomicity/java/Race.java
// Layer 1 - Why `counter++` is not atomic, Java version.
// A plain `long` field has no atomicity or visibility guarantees across
// threads under the Java Memory Model unless it's `volatile` (visibility
// only, not atomicity for read-modify-write) or you use
// java.util.concurrent.atomic / synchronized (both). Java's real threads
// and its JIT don't get to quietly optimize this race away the way Rust's
// and (likely) C++'s LLVM-based toolchains might -- see this experiment's
// result against those two.
import java.util.concurrent.atomic.AtomicLong;

public class Race {
    static final int THREADS = 8;
    static final long INCREMENTS = 300_000;

    static long counter = 0;

    static long runUnsafe() throws InterruptedException {
        counter = 0;
        Thread[] threads = new Thread[THREADS];
        for (int i = 0; i < THREADS; i++) {
            threads[i] = new Thread(() -> {
                for (long j = 0; j < INCREMENTS; j++) {
                    counter++; // racy: not synchronized
                }
            });
            threads[i].start();
        }
        for (Thread t : threads) t.join();
        return counter;
    }

    static long runSynchronized() throws InterruptedException {
        counter = 0;
        Object lock = new Object();
        Thread[] threads = new Thread[THREADS];
        for (int i = 0; i < THREADS; i++) {
            threads[i] = new Thread(() -> {
                for (long j = 0; j < INCREMENTS; j++) {
                    synchronized (lock) {
                        counter++;
                    }
                }
            });
            threads[i].start();
        }
        for (Thread t : threads) t.join();
        return counter;
    }

    static long runAtomic() throws InterruptedException {
        AtomicLong atomicCounter = new AtomicLong(0);
        Thread[] threads = new Thread[THREADS];
        for (int i = 0; i < THREADS; i++) {
            threads[i] = new Thread(() -> {
                for (long j = 0; j < INCREMENTS; j++) {
                    atomicCounter.incrementAndGet();
                }
            });
            threads[i].start();
        }
        for (Thread t : threads) t.join();
        return atomicCounter.get();
    }

    public static void main(String[] args) throws InterruptedException {
        long expected = (long) THREADS * INCREMENTS;
        long unsafeResult = runUnsafe();
        long syncResult = runSynchronized();
        long atomicResult = runAtomic();
        System.out.printf("expected:                %d%n", expected);
        System.out.printf("unsafe (counter++):      %d  (lost %d)%n", unsafeResult, expected - unsafeResult);
        System.out.printf("safe (synchronized):     %d%n", syncResult);
        System.out.printf("safe (AtomicLong):       %d%n", atomicResult);
    }
}
